File: identity-mdoc/src/main/java/com/android/identity/mdoc/request/DeviceRequestGenerator.java
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity.mdoc.request;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.identity.cbor.ArrayBuilder;
import com.android.identity.cbor.Bstr;
import com.android.identity.cbor.Cbor;
import com.android.identity.cbor.CborArray;
import com.android.identity.cbor.CborBuilder;
import com.android.identity.cbor.CborMap;
import com.android.identity.cbor.DataItem;
import com.android.identity.cbor.DataItemExtensionsKt;
import com.android.identity.cbor.MapBuilder;
import com.android.identity.cbor.RawCbor;
import com.android.identity.cbor.Tagged;
import com.android.identity.cose.Cose;
import com.android.identity.cose.CoseLabel;
import com.android.identity.cose.CoseNumberLabel;
import com.android.identity.crypto.Algorithm;
import com.android.identity.crypto.CertificateChain;
import com.android.identity.crypto.EcPrivateKey;

import java.util.Map;

/**
 * Helper class for building <code>DeviceRequest</code> <a href="http://cbor.io/">CBOR</a>
 * as specified in <em>ISO/IEC 18013-5</em> section 8.3 <em>Device Retrieval</em>.
 *
 * <p>This class supports requesting data for multiple documents in a single presentation.
 */
public final class DeviceRequestGenerator {
    private static final String TAG = "DeviceRequestGenerator";

    private final ArrayBuilder<CborBuilder> mDocRequestsBuilder;
    private byte[] mEncodedSessionTranscript;

    /**
     * Constructs a new {@link DeviceRequestGenerator}.
     */
    public DeviceRequestGenerator() {
        mDocRequestsBuilder = CborArray.Companion.builder();
    }

    /**
     * Sets the bytes of the <code>SessionTranscript</code> CBOR.
     *
     * This must be called if any of the document requests use reader authentication.
     *
     * @param encodedSessionTranscript the bytes of <code>SessionTranscript</code>.
     * @return the <code>DeviceRequestGenerator</code>.
     */
    @NonNull
    public DeviceRequestGenerator setSessionTranscript(
            @NonNull byte[] encodedSessionTranscript) {
        mEncodedSessionTranscript = encodedSessionTranscript;
        return this;
    }

    /**
     * Adds a request for a document and which data elements to request.
     *
     * @param docType                   the document type.
     * @param itemsToRequest            the items to request as a map of namespaces into data
     *                                  element
     *                                  names into the intent-to-retain for each data element.
     * @param requestInfo               <code>null</code> or additional information provided.
     *                                  This is
     *                                  a map from keys and the values must be valid
     *                                  <a href="http://cbor.io/">CBOR</a>.
     * @param readerKey                 <code>null</code> if not signing the request, otherwise a
     *                                  {@link EcPrivateKey} to be used for signing the request.
     * @param signatureAlgorithm        {@link Algorithm#UNSET} if <code>readerKey</code> is
     *                                  <code>null</code>, otherwise the signature algorithm to use.
     * @param readerKeyCertificateChain <code>null</code> if <code>readerKey</code> is
     *                                  <code>null</code>, otherwise a chain of X.509
     *                                  certificates for <code>readerKey</code>.
     * @return the <code>DeviceRequestGenerator</code>.
     */
    @NonNull
    public DeviceRequestGenerator addDocumentRequest(@NonNull String docType,
            @NonNull Map<String, Map<String, Boolean>> itemsToRequest,
            @Nullable Map<String, byte[]> requestInfo,
            @Nullable EcPrivateKey readerKey,
            Algorithm signatureAlgorithm,
            @Nullable CertificateChain readerKeyCertificateChain) {

        // TODO: Add variant that can sign with SecureArea readerKey

        MapBuilder<CborBuilder> nsBuilder = CborMap.Companion.builder();

        for (String namespaceName : itemsToRequest.keySet()) {
            Map<String, Boolean> innerMap = itemsToRequest.get(namespaceName);
            MapBuilder<MapBuilder<CborBuilder>> elemBuilder = nsBuilder.putMap(namespaceName);
            for (String elemName : innerMap.keySet()) {
                boolean intentToRetain = innerMap.get(elemName).booleanValue();
                elemBuilder.put(elemName, intentToRetain);
            }
            elemBuilder.end();
        }
        nsBuilder.end();

        MapBuilder<CborBuilder> irMapBuilder = CborMap.Companion.builder();
        irMapBuilder.put("docType", docType);
        irMapBuilder.put("nameSpaces", nsBuilder.end().build());
        if (requestInfo != null) {
            MapBuilder<MapBuilder<CborBuilder>> riBuilder = irMapBuilder.putMap("requestInfo");
            for (String key : requestInfo.keySet()) {
                byte[] value = requestInfo.get(key);
                DataItem valueDataItem = Cbor.decode(value);
                riBuilder.put(key, valueDataItem);
            }
            riBuilder.end();
        }
        irMapBuilder.end();
        byte[] encodedItemsRequest = Cbor.encode(irMapBuilder.end().build());

        DataItem itemsRequestBytesDataItem = new Tagged(24, new Bstr(encodedItemsRequest));

        DataItem readerAuth = null;
        if (readerKey != null) {
            if (readerKeyCertificateChain == null) {
                throw new IllegalArgumentException("readerKey is provided but no cert chain");
            }
            if (mEncodedSessionTranscript == null) {
                throw new IllegalStateException("sessionTranscript has not been set");
            }

            byte[] encodedReaderAuthentication = Cbor.encode(
                    CborArray.Companion.builder()
                    .add("ReaderAuthentication")
                    .add(new RawCbor(mEncodedSessionTranscript))
                    .add(itemsRequestBytesDataItem)
                    .end()
                    .build());

            byte[] readerAuthenticationBytes =
                    Cbor.encode(new Tagged(24, new Bstr(encodedReaderAuthentication)));

            Map<CoseLabel, DataItem> protectedHeaders = Map.of(
                    new CoseNumberLabel(Cose.COSE_LABEL_ALG),
                    DataItemExtensionsKt.getToDataItem(signatureAlgorithm.getCoseAlgorithmIdentifier())
            );
            Map<CoseLabel, DataItem> unprotectedHeaders = Map.of(
                    new CoseNumberLabel(Cose.COSE_LABEL_X5CHAIN),
                    readerKeyCertificateChain.getDataItem()
            );

            readerAuth = Cose.coseSign1Sign(
                    readerKey,
                    readerAuthenticationBytes,
                    false,
                    signatureAlgorithm,
                    protectedHeaders,
                    unprotectedHeaders
            ).getToDataItem();
        }

        MapBuilder<CborBuilder> mapBuilder = CborMap.Companion.builder();
        mapBuilder.put("itemsRequest", itemsRequestBytesDataItem);
        if (readerAuth != null) {
            mapBuilder.put("readerAuth", readerAuth);
        }
        DataItem docRequest = mapBuilder.end().build();

        mDocRequestsBuilder.add(docRequest);
        return this;
    }

    /**
     * Builds the <code>DeviceRequest</code> CBOR.
     *
     * @return the bytes of <code>DeviceRequest</code> CBOR.
     */
    @NonNull
    public byte[] generate() {
        return Cbor.encode(CborMap.Companion.builder()
                .put("version", "1.0")
                .put("docRequests", mDocRequestsBuilder.end().build())
                .end()
                .build());
    }

}

File: multipaz/src/commonMain/kotlin/org/multipaz/mdoc/request/DeviceRequestGenerator.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.multipaz.mdoc.request

import org.multipaz.cbor.Bstr
import org.multipaz.cbor.Cbor
import org.multipaz.cbor.CborArray
import org.multipaz.cbor.CborMap
import org.multipaz.cbor.DataItem
import org.multipaz.cbor.RawCbor
import org.multipaz.cbor.Tagged
import org.multipaz.cbor.addCborMap
import org.multipaz.cbor.buildCborArray
import org.multipaz.cbor.buildCborMap
import org.multipaz.cbor.putCborArray
import org.multipaz.cbor.putCborMap
import org.multipaz.cbor.toDataItem
import org.multipaz.cose.Cose
import org.multipaz.cose.Cose.coseSign1Sign
import org.multipaz.cose.CoseLabel
import org.multipaz.cose.CoseNumberLabel
import org.multipaz.crypto.Algorithm
import org.multipaz.crypto.X509CertChain
import org.multipaz.crypto.EcPrivateKey
import org.multipaz.mdoc.zkp.ZkSystemSpec
import org.multipaz.securearea.KeyUnlockData
import org.multipaz.securearea.SecureArea

/**
 * Helper class for building `DeviceRequest` [CBOR](http://cbor.io/)
 * as specified in ISO/IEC 18013-5:2021 section 8.3 Device Retrieval.
 *
 * This class supports requesting data for multiple documents in a single presentation.
 *
 * @param encodedSessionTranscript the bytes of `SessionTranscript`.
 */
class DeviceRequestGenerator(
    val encodedSessionTranscript: ByteArray
) {
    private val docRequestsBuilder = CborArray.builder()

    /**
     * Adds a request for a document and which data elements to request.
     *
     * This variant signs with a key in a [SecureArea].
     *
     * @param docType the document type.
     * @param itemsToRequest the items to request as a map of namespaces into data
     * element names into the intent-to-retain for each data element.
     * @param requestInfo null or additional information provided. This is
     * a map from keys and the values must be valid CBOR
     * @param readerKeySecureArea the [SecureArea] that holds the key to sign with.
     * @param readerKeyAlias the alias for the key to sign with
     * @param readerKeyCertificateChain the certification for the reader key.
     * @param keyUnlockData a [KeyUnlockData] for unlocking the key in the [SecureArea].
     * @return the [DeviceRequestGenerator].
     */
    suspend fun addDocumentRequest(
        docType: String,
        itemsToRequest: Map<String, Map<String, Boolean>>,
        requestInfo: Map<String, ByteArray>?,
        readerKeySecureArea: SecureArea,
        readerKeyAlias: String,
        readerKeyCertificateChain: X509CertChain,
        keyUnlockData: KeyUnlockData?,
    ): DeviceRequestGenerator = apply {
        val encodedItemsRequest = Cbor.encode(
            buildCborMap {
                put("docType", docType)
                putCborMap("nameSpaces") {
                    for ((namespaceName, innerMap) in itemsToRequest) {
                        putCborMap(namespaceName) {
                            for ((elemName, intentToRetain) in innerMap) {
                                put(elemName, intentToRetain)
                            }
                        }
                    }
                }
                if (requestInfo != null) {
                    putCborMap("requestInfo") {
                        for ((key, value) in requestInfo) {
                            put(key, RawCbor(value))
                        }
                    }
                }
            }
        )
        val itemsRequestBytesDataItem = Tagged(Tagged.ENCODED_CBOR, Bstr(encodedItemsRequest))

        val keyInfo = readerKeySecureArea.getKeyInfo(readerKeyAlias)

        val encodedReaderAuthentication = Cbor.encode(
            buildCborArray {
                add("ReaderAuthentication")
                add(RawCbor(encodedSessionTranscript))
                add(itemsRequestBytesDataItem)
            }
        )
        val readerAuthenticationBytes =
            Cbor.encode(Tagged(24, Bstr(encodedReaderAuthentication)))
        val protectedHeaders = mapOf<CoseLabel, DataItem>(
            Pair(
                CoseNumberLabel(Cose.COSE_LABEL_ALG),
                keyInfo.algorithm.coseAlgorithmIdentifier!!.toDataItem()
            )
        )
        val unprotectedHeaders = mapOf<CoseLabel, DataItem>(
            Pair(
                CoseNumberLabel(Cose.COSE_LABEL_X5CHAIN),
                readerKeyCertificateChain.toDataItem()
            )
        )
        val readerAuth = coseSign1Sign(
            secureArea = readerKeySecureArea,
            alias = readerKeyAlias,
            message = readerAuthenticationBytes,
            includeMessageInPayload = false,
            protectedHeaders = protectedHeaders,
            unprotectedHeaders = unprotectedHeaders,
            keyUnlockData = keyUnlockData
        ).toDataItem()

        docRequestsBuilder.add(
            buildCborMap {
                put("itemsRequest", itemsRequestBytesDataItem)
                put("readerAuth", readerAuth)
            }
        )
    }

    /**
     * Adds a request for a document and which data elements to request.
     *
     * @param docType the document type.
     * @param itemsToRequest the items to request as a map of namespaces into data
     * element names into the intent-to-retain for each data element.
     * @param requestInfo null or additional information provided. This is
     * a map from keys and the values must be valid CBOR
     * @param readerKey `null` if not signing the request, otherwise a [EcPrivateKey].
     * @param signatureAlgorithm [Algorithm.UNSET] if readerKey is null, otherwise algorithm to use.
     * @param readerKeyCertificateChain null if readerKey is null, otherwise the certificate chain
     * for readerKey.
     * @return the [DeviceRequestGenerator].
     */
    fun addDocumentRequest(
        docType: String,
        itemsToRequest: Map<String, Map<String, Boolean>>,
        requestInfo: Map<String, ByteArray>?,
        readerKey: EcPrivateKey?,
        signatureAlgorithm: Algorithm,
        readerKeyCertificateChain: X509CertChain?,
        zkSystemSpecs: List<ZkSystemSpec> = emptyList()
    ): DeviceRequestGenerator = apply {
        val requestInfoMutableMap = requestInfo?.toMutableMap() ?: mutableMapOf()
        if (zkSystemSpecs.isNotEmpty()) {
            requestInfoMutableMap["zkRequest"] = Cbor.encode(
                buildCborMap {
                    putCborArray("systemSpecs") {
                        for (zkSpec in zkSystemSpecs) {
                            addCborMap {
                                put("id", zkSpec.id)
                                put("system", zkSpec.system)
                                if (zkSpec.params.isNotEmpty()) {
                                    putCborMap("params") {
                                        for (param in zkSpec.params) {
                                            put(param.key, param.value.toDataItem())
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            )
        }

        val encodedItemsRequest = Cbor.encode(
            buildCborMap {
                put("docType", docType)
                putCborMap("nameSpaces") {
                    for ((namespaceName, innerMap) in itemsToRequest) {
                        putCborMap(namespaceName) {
                            for ((elemName, intentToRetain) in innerMap) {
                                put(elemName, intentToRetain)
                            }
                        }
                    }
                }
                if (requestInfoMutableMap.isNotEmpty()) {
                    putCborMap("requestInfo") {
                        for ((key, value) in requestInfoMutableMap) {
                            put(key, RawCbor(value))
                        }
                    }
                }
            }
        )
        val itemsRequestBytesDataItem = Tagged(Tagged.ENCODED_CBOR, Bstr(encodedItemsRequest))

        var readerAuth: DataItem? = null
        if (readerKey != null) {
            requireNotNull(readerKeyCertificateChain) { "readerKey is provided but no cert chain" }
            val encodedReaderAuthentication = Cbor.encode(
                buildCborArray {
                    add("ReaderAuthentication")
                    add(RawCbor(encodedSessionTranscript))
                    add(itemsRequestBytesDataItem)
                }
            )
            val readerAuthenticationBytes =
                Cbor.encode(Tagged(24, Bstr(encodedReaderAuthentication)))
            val protectedHeaders = mapOf<CoseLabel, DataItem>(
                Pair(
                    CoseNumberLabel(Cose.COSE_LABEL_ALG),
                    signatureAlgorithm.coseAlgorithmIdentifier!!.toDataItem()
                )
            )
            val unprotectedHeaders = mapOf<CoseLabel, DataItem>(
                Pair(
                    CoseNumberLabel(Cose.COSE_LABEL_X5CHAIN),
                    readerKeyCertificateChain.toDataItem()
                )
            )
            readerAuth = coseSign1Sign(
                key = readerKey,
                dataToSign = readerAuthenticationBytes,
                includeDataInPayload = false,
                signatureAlgorithm = signatureAlgorithm,
                protectedHeaders = protectedHeaders,
                unprotectedHeaders = unprotectedHeaders
            ).toDataItem()
        }

        docRequestsBuilder.add(
            buildCborMap {
                put("itemsRequest", itemsRequestBytesDataItem)
                if (readerAuth != null) {
                    put("readerAuth", readerAuth)
                }
            }
        )
    }

    /**
     * Builds the `DeviceRequest` CBOR.
     *
     * @return the bytes of `DeviceRequest` CBOR.
     */
    fun generate(): ByteArray = Cbor.encode(
        buildCborMap {
            put("version", "1.0")
            put("docRequests", docRequestsBuilder.end().build())
        }
    )
}
